src/js/common/containers/DowntimeNotification.jsx

```diff
@@ -0,0 +1,244 @@
+    determineStatus: PropTypes.func,
+    getScheduledDowntime: PropTypes.func.isRequired,
+    isReady: PropTypes.bool,
+    loadingIndicator: PropTypes.node,
+    render: PropTypes.func,
+    scheduledDowntime: PropTypes.arrayOf(
+      PropTypes.shape({
+        service: PropTypes.string,
+        description: PropTypes.string,
+        startTime: PropTypes.instanceOf(Date),
+        endTime: PropTypes.instanceOf(Date)
+      })
+    )
+  };
+
+  static defaultProps = {
+    appTitle: 'application'
+  };
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      modalDismissed: false,
+      cache: {}
+    };
+  }
+
+  componentWillMount() {
+    this.props.getScheduledDowntime();
+  }
+
+  // This is here just for caching the result of calculateDowntime, so that it isn't run every time a prop changes.
+  // Currently, this component should only do its calculations once, because it would be weird if in the middle of filling
+  // out a form a modal or alert appears alerting them of downtime, or if it just shuts down the app because we entered
+  // scheduled downtime.
+  componentWillReceiveProps(newProps) {
+    const firstLoad = newProps.isReady && !this.props.isReady;
+    if (firstLoad) {
+      const downtimeMap = this.calculateDowntime(newProps.dependencies, newProps.scheduledDowntime);
+      const status = this.determineStatus(downtimeMap);
+      const downtimeWindow = this.getDowntimeWindow(downtimeMap.get(status));
+      const cache = { downtimeMap, status, downtimeWindow };
+      this.setState({ cache });
+    }
+  }
+
+  shouldComponentUpdate(nextProps, nextState) {
+    if (!nextProps.isReady) return false;
+    if (nextProps.isReady && !this.props.isReady) return true;
+
+    return nextState.cache.status !== serviceStatus.down;
+  }
+
+  getStatusForDowntime(downtime, now = moment()) {
+    const inclusive = '[]';
+    if (now.isBetween(downtime.startTime, downtime.endTime, inclusive)) return serviceStatus.down;
+    if (now.add(1, 'hour').isAfter(downtime.startTime)) return serviceStatus.downtimeApproaching;
+    return serviceStatus.ok;
+  }
+
+  // This is used to calculate the timeframe of the current or impending downtime.
+  // To keep things simple, it finds the downtime with the soonest start time and uses those
+  // start/end times.
+  getDowntimeWindow(downtimes) {
+    // Cycle through the array of downtimes and find the downtime with the soonest start time.
+    const soonestDowntime = downtimes.reduce((currentSoonest, downtime) => {
+      if (!currentSoonest) return downtime;
+
+      const isEarlier = moment(downtime.startTime).isBefore(currentSoonest.startTime);
+      return isEarlier ? downtime : currentSoonest;
+    }, null);
+
+    // Return the startTime/endTime as moment objects for the soonest downtime.
+    return {
+      startTime: soonestDowntime && moment(soonestDowntime.startTime),
+      endTime: soonestDowntime && moment(soonestDowntime.endTime)
```

Comment:
Maybe splitting hairs here because this may never happen, but if we've got the following downtimes
```js
downtimes = [
  { startTime: '12:00', endTime: '13:00' },
  {  endTime: '15:00' }
]
```
I think this will return
```js
{
  startTime: moment(undefined), // resolves to the current time
  endTime: moment('15:00')
}
```
when the current time is < 12:00. This might be the intended behavior, but I wanted to point it out.

Reply:
Actually it's only looking at the `startTime` now and grabbing the corresponding `endTime` off that. So that should return 

```
{
  startTime: moment('12:00'), // resolves to the current time
  endTime: moment('13:00')
}
```
But that brings up a good point, because the endTime can be omitted in which case we shouldn't pass it into moment. I just published a patch for that case.

Reply:
Oh I see what you mean. The `startTime` is required though, so that shouldn't ever happen. `endTime` though is optional, which isn't clear here. I just published a patch for this.